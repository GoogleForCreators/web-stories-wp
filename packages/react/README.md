# React

Generic package that re-exports parts of `react` and `react-dom` alongside some other helpful utils and hooks.
