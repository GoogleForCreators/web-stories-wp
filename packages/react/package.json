{
  "name": "@googleforcreators/react",
  "description": "A tiny wrapper package that re-exports React and provides additional utilities and custom hooks.",
  "private": false,
  "version": "0.1.202204141511",
  "author": {
    "name": "Google",
    "url": "https://creators.google/"
  },
  "license": "Apache-2.0",
  "keywords": [
    "web stories"
  ],
  "homepage": "https://github.com/GoogleForCreators/web-stories-wp/blob/main/packages/react/README.md",
  "repository": {
    "type": "git",
    "url": "https://github.com/GoogleForCreators/web-stories-wp.git",
    "directory": "packages/react"
  },
  "bugs": {
    "url": "https://github.com/googleforcreators/web-stories-wp/issues"
  },
  "engines": {
    "node": ">= 12 || >= 14 || >= 16",
    "npm": ">= 7.3"
  },
  "customExports": {
    ".": {
      "default": "./src/index.js"
    }
  },
  "main": "dist/index.js",
  "module": "dist-module/index.js",
  "source": "src/index.js",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "prop-types": "^15.8.1",
    "resize-observer-polyfill": "^1.5.1",
    "shallow-equal": "^1.2.1",
    "use-context-selector": "^1.2.11",
    "use-debounce": "^7.0.1"
  },
  "devDependencies": {
    "@testing-library/react": "^12.1.5",
    "@testing-library/react-hooks": "^8.0.0"
  },
  "peerDependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  }
}
