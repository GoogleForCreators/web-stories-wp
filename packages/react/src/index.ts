/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { useDebouncedCallback } from 'use-debounce';
export { default as useReduction } from 'use-reduction';

export * from './context';
export * from './react';
export * from './react-dom';

export { default as useBatchingCallback } from './useBatchingCallback';
export { default as useCombinedRefs } from './useCombinedRefs';
export { default as useContextSelector } from './useContextSelector';
export { default as useFocusOut } from './useFocusOut';
export { default as usePrevious } from './usePrevious';
export { default as useResizeEffect } from './useResizeEffect';
export { default as renderToStaticMarkup } from './renderToStaticMarkup';
export { default as useUnmount } from './useUnmount';
export { default as useWhyDidYouUpdate } from './useWhyDidYouUpdate';
export { default as useInitializedValue } from './useInitializedValue';
export { default as shallowEqual } from './shallowEqual';
