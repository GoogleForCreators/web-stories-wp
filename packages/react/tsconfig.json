{
  "extends": "../../tsconfig.shared.json",
  "compilerOptions": {
    "rootDir": "src",
    "declarationDir": "dist-types"
  },
  "include": ["src/context.ts", "src/react.ts", "src/react-dom.ts"]
}
