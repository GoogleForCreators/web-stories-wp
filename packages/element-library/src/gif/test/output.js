/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { renderToStaticMarkup } from '@googleforcreators/react';

/**
 * Internal dependencies
 */
import GifOutput from '../output';

describe('Gif Output', () => {
  const baseProps = {
    element: {
      id: '048426c8-69ae-4e04-80f1-8f3fd4434261',
      type: 'gif',
      opacity: 100,
      x: 65,
      y: 196,
      width: 281,
      height: 223,
      scale: 100,
      rotationAngle: 0,
      resource: {
        type: 'gif',
        mimeType: 'image/gif',
        creationDate: '2016-02-04T18:16:22Z',
        src: 'https://c.tenor.com/4F2m7BWP6KYAAAAC/flying-kiss-muah.gif',
        width: 281,
        height: 223,
        alt: '',
        attribution: {
          author: [],
          registerUsageUrl:
            'https://media3p.googleapis.com/v1/media:registerUsage?token=AX7RMSdePGQBB3c/QAOBJ20QC%2BZNp2A549gSosisUYOjIC71nkvySPH5yj%2BqDBOVBmmFZ89azzUAN9x2GjkNbq3OXauUMho%3D',
        },
        output: {
          mimeType: 'video/mp4',
          src: 'https://c.tenor.com/4F2m7BWP6KYAAAPo/flying-kiss-muah.mp4',
        },
      },
    },
    box: {
      x: 15.7767,
      y: 31.71521,
      width: 68.20388,
      height: 36.08414,
      rotationAngle: 0,
    },
  };

  it('should produce an AMP video with autoplay, no controls, no audio, and loop', async () => {
    const output = <GifOutput {...baseProps} />;
    const outputStr = renderToStaticMarkup(output);
    await expect(outputStr).toStrictEqual(
      expect.stringMatching(
        'src="https://c.tenor.com/4F2m7BWP6KYAAAPo/flying-kiss-muah.mp4"'
      )
    );
    await expect(outputStr).toMatchSnapshot();
  });

  it('should include poster image if available', async () => {
    const newProps = { ...baseProps };
    newProps.element.resource.poster =
      'https://c.tenor.com/4F2m7BWP6KYAAAAC/flying-kiss-muah-poster.png';
    const output = <GifOutput {...newProps} />;
    const outputStr = renderToStaticMarkup(output);
    await expect(outputStr).toStrictEqual(
      expect.stringMatching(
        'src="https://c.tenor.com/4F2m7BWP6KYAAAPo/flying-kiss-muah.mp4"'
      )
    );
    await expect(outputStr).toMatchSnapshot();
  });

  it('should remove blob URLs', async () => {
    const props = {
      ...baseProps,
      element: {
        ...baseProps.element,
        resource: {
          ...baseProps.element.resource,
          output: {
            ...baseProps.element.resource.output,
            src: 'blob:https://example.com/ecee4374-8f8a-4210-8f2d-9c5f8d6a6c5a',
          },
        },
      },
    };
    const output = <GifOutput {...props} />;
    const outputStr = renderToStaticMarkup(output);
    await expect(outputStr).not.toStrictEqual(expect.stringMatching('blob:'));
  });

  describe('AMP validation', () => {
    jest.retryTimes(3, { logErrorsBeforeRetry: true });

    it('should produce valid AMP output', async () => {
      await expect(<GifOutput {...baseProps} />).toBeValidAMPStoryElement();
    });

    it('should produce valid AMP output with poster', async () => {
      const newProps = { ...baseProps };
      newProps.element.resource.poster =
        'https://c.tenor.com/4F2m7BWP6KYAAAAC/flying-kiss-muah-poster.png';
      await expect(<GifOutput {...newProps} />).toBeValidAMPStoryElement();
    });
  });
});
