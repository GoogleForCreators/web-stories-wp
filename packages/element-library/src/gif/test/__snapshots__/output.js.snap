// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gif Output should include poster image if available 1`] = `"<div style="position:absolute;width:100%;height:100%;left:0%;top:0%" data-leaf-element="true"><amp-video id="el-048426c8-69ae-4e04-80f1-8f3fd4434261-media" autoplay="autoplay" loop="loop" noaudio="noaudio" poster="https://c.tenor.com/4F2m7BWP6KYAAAAC/flying-kiss-muah-poster.png" layout="fill" alt=""><source type="video/mp4" src="https://c.tenor.com/4F2m7BWP6KYAAAPo/flying-kiss-muah.mp4"/></amp-video></div>"`;

exports[`Gif Output should produce an AMP video with autoplay, no controls, no audio, and loop 1`] = `"<div style="position:absolute;width:100%;height:100%;left:0%;top:0%" data-leaf-element="true"><amp-video id="el-048426c8-69ae-4e04-80f1-8f3fd4434261-media" autoplay="autoplay" loop="loop" noaudio="noaudio" poster="" layout="fill" alt=""><source type="video/mp4" src="https://c.tenor.com/4F2m7BWP6KYAAAPo/flying-kiss-muah.mp4"/></amp-video></div>"`;
